Metric SDK Accumulator requirements and details (#880)
open-telemetry/opentelemetry-specification: specification/metrics/sdk.md

language implementors to follow in any approach to implementing the
OpenTelemetry API.

In the second part, the architectural details of a model SDK
implementation are described, using the OpenTelemetry-Go Metric SDK as
an example.  This is meant to show how the requirements can be
implemented, as guidance to implementors, not to mandate an exact
reproduction of the model architecture across languages.
OpenTelemetry SDKs are of course free to use any idiomatic approach.

## Expectations

This document assumes you are familiar with the OpenTelemetry [Metrics
API specification](api.md).

The SDK implementors are expected to follow the best practices for the
language and runtime environment when implementing the OpenTelemetry
API.  Implementors SHOULD follow the general prescriptions on safety
and performance given in [OpenTelemetry library
guidelines](../library-guidelines.md).

## SDK Terminology

The Metrics SDK provides an implementation of the Metrics API, which
has terms that it uses to refer to the SDK:

- **Meter**: An interface that supplies the OpenTelemetry Metric API, bound together with Resources and the Instrumentation Library

> - **Provider**: An interface for obtaining **Meter** instances given the Instrumentation Library.

Reviewer: This leads me to believe that the export pipeline would be Provider -> gets meter (provider can have many meter instances) -> meter implementation has an accumulator (singleton) -> ... correct?

If provider can create many meter instances, how are they identifiable? InstrumentationInfo? Name?

Are these details too specific for the spec?

> - **Provider**: An interface for obtaining **Meter** instances given the Instrumentation Library.

Reviewer: Will there be a `Provider Detail` as well? Will configurations like `stateful` exist there?

Author: I think of `Provider` as an interface given to the user for access to the metrics API, just a getter for Meters.
I think of a configuration like `stateful` as belonging in the `Processor`. The way the Go SDK is implemented, there are two related properties having to do with state. These configurations relate to long-term state, which resides in the Processor, whereas the Accumulator is only responsible for short-term state. The two properties are:

1. Should the processor retain memory about values where the export kind (cumulative, delta) disagrees with the instrument kind
2. Should the processor retain memory about values _just because_ the exporter wants it.

The first property is true when a Counter exports with cumulative aggregation temporality or when a SumObserver exports with delta temporality (*).

The second property is true when the exporter wants the processor to remember values, even when the first property does not apply. This is true for SumObservers exporting cumulative values in Prometheus, for example: the exporter is expected to continue reporting values, even if the application stops updating the instrument.

(*) actually this case "SumObserver exporting as delta" has dubious value, but it's implemented in Go to show how the Processor could do this.

Reviewer: Alright. I am not too sure how configuration will be done on the `Processor` side (currently in Python the user never interacts with it), but might get more clarity once the `Processor` portion is worked on.

> - **Provider**: An interface for obtaining **Meter** instances given the Instrumentation Library.

Reviewer 2: Is this Provider same as `MeterProvider` https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/metrics/api.md#meter-provider?
Also, its the equivalent of `TracerProvider` ?

Author: Yes, I clarified this (spelled out `MeterProvider`).

These terms are useful to describe the boundary between the API and
the SDK, but they are API-level constructs.  We can use API-level
terms to describe the boundary between the SDK and the API, but from
the API's perspective the SDK is opaque and without structure.

This document specifies the internal structure of the default

> OpenTelemetry SDK in terms of its constituent parts, with terminologuy

Reviewer: ```suggestion
OpenTelemetry SDK in terms of its constituent parts, with terminology
```

Author: Fixed.

chosen to explain each part's role in exporting metric data from the
input (API-level events) to the output (a metric exposition format).
We use the term **Export Pipeline** as a functional description of the
SDK at this level.

There are three major components of the export pipeline that data
flows through, in order:

> 1. **Accumulator**: Receives metric events from the API through Instruments, computes one Accumulation per active Instrument and Label Set pair

Reviewer: Would it be good to clarify what `Accumulation` is?

# Metrics SDK

Note: This document assumes you are familiar with the (Metrics
API)[api.md] specification.  Note that the examples below are copied
from the current OpenTelemetry-Go SDK:

> TODO: TOC

Reviewer 3: ```suggestion
TODO: TOC
TODO: Write a list/description of the functional requirements of the SDK and put it above the model implementation description.
```

> from the current OpenTelemetry-Go SDK:

Reviewer 4: Is it worth to specifically call out the obvious: that other languages can use equivalent but syntactically different API that is idiomatic for the language?

Author: I've rewritten the first couple of paragraphs so that this point is more emphasized up front. Really this is what @jkwatson was asking for too. The requirements are separated from the model architecture details.

There are three major components of the Metrics SDK that data flows
through, in order:

> 1. **Accumulator**: Receives metric events from the API through Instruments, computes one Accumulation per active Instrument and Label Set pair

Reviewer 3: This seems to preclude an OO model where the individual instrument instances accumulate their own data. Is such a centralized component necessary? Is there only one instance per SDK? per Meter?

Author: How does `RecordBatch` work? Isn't there some coordination across instruments necessary?

Reviewer 3: Currently, it just delegates off to the individual instruments (so...it doesn't really batch at this point):  https://github.com/open-telemetry/opentelemetry-java/blob/26597266dfdaf977b7c144d0bc4cbb459bd4a62d/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/BatchRecorderSdk.java

Author: I've added more text in 2ab88b9 to explicitly state that the diagram is an abstraction but that we're aiming to standardize and require these terms.

Note that Accumulator is not a class that I expect to be configured by users. The reason why I think it should have a standard name is that we're going to write requirements about how it performs its specific task, particularly about it being able to forget label sets that are no longer in use.  This is not a requirement for the SDK as a whole, since some Processors will be designed explicitly to not forget label sets that are no longer in use. Making this an Accumulator requirement means that it is possible to implement a stateless client as long as the Processor and Exporter support that too.

Author: >> How does RecordBatch work?

> Currently, it just delegates off to the individual instruments (so...it doesn't really batch at this point)

FWIW, I think this is fine. The point of the RecordBatch API is both a semantic hint (these are atomic, you might try not to allow partial collection--not a requirement) and an optimization (you can process a label set once and use it for each measurement in the batch--I'd call this a requirement).

- **Aggregation**: the result of aggregating one or more events by a specific aggregator
- **AggregationKind**: describes the kind of read API the Aggregation supports (e.g., Sum)
- **Accumulation**: consists of Instrument, Label Set, Resource, and Aggregator snapshot
- **Controller**: coordinates the Accumulator, Processor, and Exporter components in an export pipeline
- **ExportKind**: one of Delta, Cumulative, or Pass-Through
- **ExportKindSelector**: chooses which ExportKind to use for a metric instrument.
- **ExportRecord**: consists of Instrument, Label Set, Resource, Timestamp(s), and Aggregation
- **ExportRecordSet**: a set of export records

## Dataflow Diagram

![Metrics SDK Design Diagram](img/metrics-sdk.png)

## Accumulator: Meter Implementation

The Accumulator is the first component an OpenTelemetry Metric export
pipeline, implementing the front-line [`Meter`

> interface]((api.md#meter-interface)).

Reviewer 5: ```suggestion
interface](api.md#meter-interface).
```

```
// NewAccumulator constructs a new Accumulator for the given
// Processor and options.
func NewAccumulator(processor export.Processor, opts ...Option) *Accumulator
```

> The Accumulator MUST provide the option to associate a

Reviewer 3: I wouldn't put MUST/SHOULD directives in this. It's a model implementation, so it shouldn't prescribe at this level.

Author: I think the goal of this document is both to prescribe requirements for the SDK and illustrate them with a model implementation. I believe the think you call `MeterSDK` is in Java is the same as `Accumulator` (and would recommend renaming it, since the complete SDK includes more than just that component), and that the `Accumulator` MUST support resources. Does this match your understanding?

Reviewer 3: well, the MeterSdk in java doesn't do these things, from the definitions: "Receives metric events from the API, computes one Accumulation per active Instrument and Label Set pair". That's handled by individual instrument instances, each of which has an associated aggregator.

Author: I made the separation you asked, at least so far as the Accumulator text I had written.

[`Resource`](../resource/sdk.md) with the Accumulations that it
produces.

```
// WithResource sets the Resource configuration option of a Config.
func WithResource(res *resource.Resource) Option
```

The Accumulator is called by the Controller (see below) to coordinate
collection using a `Collect()` method.

```
// Collect traverses the list of active instruments and exports
// data.  Collect() may not be called concurrently.
//
// During the collection pass, the Processor will receive
// one Process(Accumulation) call per current aggregation.
//
// Returns the number of accumulations that were exported.
func (m *Accumulator) Collect(ctx context.Context) int
```

### Implement the SDK-level API matching the user-level Metric API

The OpenTelemetry Metric API specifies a number of instruments and
supports several calling conventions, giving it a relatively large
number of types and methods ("surface area").  The OpenTelemetry API

> SHOULD be designed and implemented following the style idioms of the

Reviewer 3: why are we prescribing things about the API design in this document?

Author: This is really about how the API meets the SDK. I think of this as an SDK requirement placed on the API, but we can try to reword this.

Reviewer 3: maybe replace the capital SHOULD with "has been", so it doesn't read like this doc is prescribing anything on the API design.

Author: Fixed, I think. I've rewritten the first few paragraphs and given this document two parts.

> source language, in such a way that the SDK must implement a

Reviewer 3: I don't understand what this is trying to say. How can the SDK both implement the API, and implement an interface narrower than the API?

Author: In Go, at least, the API has a large surface area and the SDK is not required to implement it exactly. The API is better designed to handle the surface area itself and pass through to a narrow interface for the SDK. I see this as a useful requirement, but maybe it should not be. I'm thinking of cases where a code base could adopt the SDK without adopting the API itself. In that case, having a narrow interface will make the SDK useful when the API is not being used.

For example, I'm converting a legacy code base to use the SDK without updating the instrumentation to use OpenTelemetry.

Reviewer 3: Maybe I'm misunderstanding your usage of the terms here. I thought it was the bare minimum requirement for an SDK that it implement the entire API, not a narrower version of the API, which would require a layer in between the API and the SDK to mediate this difference.

Author: I think I had been misunderstanding a few things. Hopefully I've removed the offensive bits entirely. 😀

// generic bound synchronous instrument
type BoundSyncImpl interface {

	// RecordOne captures a single synchronous metric event.
	RecordOne(ctx context.Context, number Number)

	// Unbind frees the resources associated with this bound instrument.
	Unbind()
}
```

#### Asynchronous SDK instrument

The [asynchronous SDK instrument](api.md#asynchronous-instrument-details)
supports both single-observer and batch-observer calling conventions.
The interface used for running Observer callbacks is passed at the
constructor, so there are no other API-level access methods for
asynchronous instruments.

```
// AsyncImpl is an implementation-level interface to an
// asynchronous instrument (e.g., Observer instruments).
type AsyncImpl interface {
        // InstrumentImpl provides Descriptor() and Implementation()
	InstrumentImpl
}
```

The asynchronous "runner" interface (`AsyncRunner`) passed in the
asynchronous SDK instrument constructor supports both
[single](api.md#single-instrument-observer) and
[batch](api.md#batch-observer) calling conventions.  These are
considered language-specific details.

### Instrument registration

The OpenTelemetry API SHOULD provide a wrapper for the Meter

> implementation (`MeterImpl`) that:

Reviewer 5: It wasn't clear to me what the wrapper does until I went and checked the Otel Go code. A little more here might be more clear
```suggestion
implementation (`MeterImpl`) that stores unique instruments for retrieval by their descriptor:
```

> 

Reviewer 5: It also creates an instrument if it doesn't exist yet, right?

Author: Fixed.

> - Provided the instrument definitions match, returns a unique SDK instrument
> - When the instrument definitions do not match, returns a errors and a no-op instrument.

Reviewer 5: ```suggestion
- Provided the instrument descriptors match, returns the unique SDK instrument
- When the instrument descriptors do not match, returns an error and a no-op instrument.
```

- **Aggregation**: the result of aggregating one or more events by a specific aggregator
- **AggregationKind**: describes the kind of read API the Aggregation supports (e.g., Sum)
- **Accumulation**: consists of Instrument, Label Set, Resource, and Aggregator snapshot
- **Controller**: coordinates the Accumulator, Processor, and Exporter components in an export pipeline
- **ExportKind**: one of Delta, Cumulative, or Pass-Through
- **ExportKindSelector**: chooses which ExportKind to use for a metric instrument.
- **ExportRecord**: consists of Instrument, Label Set, Resource, Timestamp(s), and Aggregation
- **ExportRecordSet**: a set of export records

## Dataflow Diagram

![Metrics SDK Design Diagram](img/metrics-sdk.png)

## Accumulator: Meter Implementation

The Accumulator is the first component an OpenTelemetry Metric export
pipeline, implementing the front-line [`Meter`
interface](api.md#meter-interface).

> ```

Reviewer 4: ```suggestion
```go
```
This should enable syntax highlighting in Github.

Author: Done.

	// synchronous instrument implementation or an error, should
	// one occur.
	NewSyncInstrument(descriptor Descriptor) (SyncImpl, error)

	// NewAsyncInstrument returns a newly constructed
	// asynchronous instrument implementation or an error, should
	// one occur.
	NewAsyncInstrument(
		descriptor Descriptor,
		runner AsyncRunner,
	) (AsyncImpl, error)
}
```

These methods cover all the necessary entry points for implementing
the OpenTelemetry Metric API.

[`RecordBatch`](api.md#recordbatch-calling-convention) is a user-level API
implemented directly by the SDK.

The two instrument constructors build [_synchronous_
and _asynchronous_](api.md#synchronous-and-asynchronous-instruments-compared)
SDK instruments.

The user is generally interested in the [Metric API `Meter`
interface](api.md#meter-interface), obtained through a [Metric API
`Provider` interface](api.md#meter-provider).  The `Meter` interface can be
constructed by wrapping the SDK `Meter` implementation:

```
// WrapMeterImpl constructs a `Meter` implementation from a
// `MeterImpl` implementation.
func WrapMeterImpl(impl MeterImpl, instrumentationName string, opts ...MeterOption) Meter
```

Optional to this method:

> - setting the version of the OpenTelemetry API in use.

Reviewer 4: Unclear.

Author: It was "instrumentation library", now reads "the instrumentation library version of the named Meter in use".

### Provide access to the instrument descriptor

The SDK instruments are the underlying implementation for
[OpenTelemetry Metric instruments](api.md#metric-instruments).  This
interface links API and the SDK:

```
// InstrumentImpl is a common interface for synchronous and
// asynchronous instruments.
type InstrumentImpl interface {
	// Descriptor returns a copy of the instrument's Descriptor.
	Descriptor() Descriptor
}
```

The API `Descriptor` method defines the instrument in terms of
API-level constructs including name, instrument kind, description, and
units of measure.  The SDK instrument MUST provide access to the
`Descriptor` that was passed to its constructor.

#### Synchronous SDK instrument

The [synchronous SDK instrument](api.md#synchronous-instrument-details)
supports both direct and bound calling conventions.

```
// SyncImpl is the implementation-level interface to a generic
// synchronous instrument (e.g., ValueRecorder and Counter instruments).
type SyncImpl interface {

>         // InstrumentImpl provides Descriptor() and Implementation()

Reviewer 4: There is no `Implementation()` declaration in `InstrumentImpl` above. Is it supposed to be there?

Author: Removed. There is such a method in Go, but it's really an implementation detail.

asynchronous SDK instrument constructor supports both
[single](api.md#single-instrument-observer) and
[batch](api.md#batch-observer) calling conventions.  These are
considered language-specific details.

### Instrument registration

The OpenTelemetry API SHOULD provide a wrapper for the Meter
implementation (`MeterImpl`) that stores unique instruments for retrieval by their descriptor:

- Create a new instrument when none with the same name exists
- Provided the instrument name exists and the descriptors match, returns the unique SDK instrument
- When the instrument descriptors do not match, returns an error and a no-op instrument.

```
// NewUniqueInstrumentMeterImpl returns a wrapped metric.MeterImpl with
// the addition of uniqueness checking.
func NewUniqueInstrumentMeterImpl(impl metric.MeterImpl) metric.MeterImpl {
```

### Concurrency expectations

Synchronous metric instruments are expected to be used concurrently.
Unless concurrency is not a feature of the source language, the SDK
Accumulator component SHOULD be designed with concurrent performance
in mind.

SDK implementations SHOULD use the best concurrency primitives

> available in the source language.  Accumulators MAY use exclusive

Reviewer 4: >SDK implementations SHOULD use the best concurrency primitives
>available in the source language.

Is this sentence necessary? Could we ever recommend to use something that is not the "best"? Perhaps qualify "best" to make the intent clear.

Author: Fixed. (Removed. this sentence.)

- **Controller**: coordinates the Accumulator, Processor, and Exporter components in an export pipeline
- **ExportKind**: one of Delta, Cumulative, or Pass-Through
- **ExportKindSelector**: chooses which ExportKind to use for a metric instrument.
- **ExportRecord**: consists of Instrument, Label Set, Resource, Timestamp(s), and Aggregation
- **ExportRecordSet**: a set of export records.

The term **SDK instrument** refers to the underlying implementation of
an instrument.

## Dataflow Diagram

![Metrics SDK Design Diagram](img/metrics-sdk.png)

## Requirements

Requirements are listed below for the major components of the
OpenTelemetry using the terminology outlined above.

### Instrument Registration

The OpenTelemetry SDK MUST ensure that an individual Meter
implementation cannot report instruments with the same name and
different instrument kind.

For synchronous instruments, duplicate registration is permitted.  The
SDK SHOULD return the same SDK instrument when correctly registered
multiple times.

For asynchronous instruments, duplicate registration is not permitted.
The SDK MUST not permit registration of multiple callbacks with the same
instrument name.

> ### Accumulator

Reviewer 3: I don't think we need to call out the components in the functional requirements. I'd rather we define what needs to be supported functionally without referencing internal implementation details. If a given SDK implementation doesn't have a component that matches this, that should be ok, I think.

If we do want to specify the internal details of the SDK at this level, then I'm ok with that, but that will probably mean a significant rewrite of the Java implementation to match the go implementation, and I don't think that's desired.

Author: I think we should have some degree of standardization regarding these component names. [In the Tracing SDK specification there is a "Span Processor"](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#span-processor). The goal is explicitly similar here, to create a common way of describing SDK components (Accumulator, Processor, Exporter) and their interactions (Collect, Process, Export) along with standard type names already written into the spec here.

Reviewer 3: So, it seems that your desire is to require an implementation model. I fear this will push out the possibility of releasing a spec-compliant Java Metrics SDK significantly. Are you ok with that? @bogdandrutu as the primary author of the current Java Metrics implementation, are you ok with this level of specification?

Author: I thought that was an objective for OpenTelemetry. I'm happy to hear how you think we should specify requirements without standard components, though.

Reviewer 3: I'd like to specify the inputs (the API, configuration, extension points) and outputs (exporter contracts). What happens in between those two feels like it should be up to the implementation. Just my opinion, though. And, my desire not to have to rewrite the Java metrics SDK, which I think this will require.

Reviewer 2: If we compare with specs for Trace, its very clear on what is the name of the thing which receives span - SpanProcessor, what is the thing which determines Span creation state  - Sampler and the name of the built-in Samplers.. etc.
I'd prefer to see similar standardization for Metrics as well.

(Yes, I understand this causes rework for all languages to do some refactoring to match the new spec)

Author: I've strengthened the text on this point. See 2ab88b9.

The Accumulator is the first component an OpenTelemetry Metric export

> pipeline, implementing the front-line [`Meter`

Author: I've written a bunch more to try and explain this use of terminology in 2ab88b9. 

"Meter" is an API-level interface, while "Accumulator" is an SDK-level function in an export pipeline.  There can be many Meters with different instrumentation library names. In the OTel-Go implementation there's a single Accumulator that maps one-to-one with `MeterImpl`, which is also a singleton underlying the `Provider` and `Meter` instances, so there's a one-to-one mapping between `Accumulator` and the "SDK" in OTel-Go.

The term Accumulator is useful just for stating requirements that specifically deal with accumulation, around locking and not having long-term memory.

I don't want there to be a requirement that there is just one Accumulator, that's an implementation detail or it's something we're not trying to specify, at least. There's been an experimental idea about support for different collection intervals, for example, and I would approach that by allocating one Accumulator for each interval. The Java SDK uses one Accumulator per instrument. This seems OK, and our goal is to add whatever requirements we think are required.

Here's a requirement I think we should add for Accumulators: State from a single instrument MUST NOT be placed in multiple accumulators, since accumulators can be Collected independently. Allowing different label sets for a single instrument to be collected on different intervals could lead to a bad user experience.

> pipeline, implementing the front-line [`Meter`

Author: Also, I can see from https://github.com/open-telemetry/opentelemetry-specification/pull/880#discussion_r478736789 that the Java SDK is organized in a way that there's some other component that allocates instruments.  This component (which for lack of a better term is "the SDK") has to manage a map of instruments, but there is still a logical Accumulator associated with each instrument.  I don't mean to make it a requirement that the Accumulator implements the Meter API, or that there's a one-to-one relationship between Accumulator and SDK, but I do mean to write requirements about the component that serves this function of accumulating measurements over short periods of time. Does this help?

> pipeline, implementing the front-line [`Meter`

Reviewer 3: It might be clearer to say that the Accumulator implements the *Instrument* APIs, rather than the Meter APIs, correct?

> pipeline, implementing the front-line [`Meter`

Author: I wrote "implementing the instrument-related APIs of the Meter interface."

> pipeline, implementing the front-line [`Meter`

Author: I still prefer to think of the Accumulator as being responsible for all the Meter APIs, the only difference being the `RecordBatch()` API. If we don't include `RecordBatch()` among the Accumulator responsibilities, then we need to assign that responsibility to some other component, and there isn't a component (or name of a component) that describes things that are the SDK but which are not one of the already-named classes. We'll end up using something like "Meter implementation" to assign some responsibilities that are very close to the Accumulator.

> pipeline, implementing the front-line [`Meter`

Reviewer: Is `RecordBatch()` not part of the Meter APIs? Although not strictly "instrument-related", wouldn't it be much simpler to just have the `Accumulator` implement all of the APIs related to recording and collection instead of having to distinguish between whether it is an instrument api or meter api?

> but I do mean to write requirements about the component that serves this function of accumulating measurements over short periods of time.

As long as this is fulfilled, whatever implementation specific details that are done should be fine correct? (one-to-one mappings of components, keeping components in memory, etc).

> pipeline, implementing the front-line [`Meter`

Reviewer 3: I guess I was trying to say that there's a difference between getting access to Instruments (which is what the Meter does, no?), and the recording of values, which is the responsibility of the Instruments, not the Meter.

If all there is is a Meter/Accumulator, why did we define all these Instruments and their functionality?

> pipeline, implementing the front-line [`Meter`

Reviewer 3: sorry for the unnecessary snark. I'm just confused about why it seems like "Meter" and "Instrument" are being conflated.

> pipeline, implementing the front-line [`Meter`

Reviewer 3: @jmacd ?

> pipeline, implementing the front-line [`Meter`

Author: @jkwatson I have taken a pass at this draft and separated parts that we think of as "the SDK" and parts that I'm trying to label "the Accumulator". I've made it clear that we're not trying to specify the exact relationship between the two, but that they serve different functions. I've added sections for the SDK itself with requirements for "Instrument Registration", "Collect() function", and "RecordBatch" (TODO). I moved this text out of the Accumulator section and I hope this addresses the core confusion you noticed.

- **Collection Interval**: the period between calls to Accumulator.Collect()
- **Label**: a key-value describing a property of the metric event
- **Label Set**: a set of key-values with unique keys
- **Measurement**: an integer or floating point number from a synchronous instrument
- **Observation**: an integer or floating point number from an asynchronous instrument.

Defined in the [Resource SDK](../resource/sdk.md) specification:

- **Resource**: a set of key-values with unique keys describing the process

> - **Instrumentation Library**: the name and version associated with a package of instrumentation.

Reviewer 6: Possibly add a link to the project glossary here?

```suggestion
- [**Instrumentation Library**](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/glossary.md#instrumentation-library): the name and version associated with a package of instrumentation.
```

These are the significant data types used in the model architecture:

- **Aggregator**: aggregates one or more measurements in a useful way

> - **Aggregator snapshot**: copy of an synchronous instrument aggregator taken during collection

Reviewer 6: For consistency should this be capitalized?

```suggestion
- **Aggregator Snapshot**: copy of an synchronous instrument aggregator taken during collection
```

to implement a top-level `Collect()` API that runs collection on one
or more Accumulator(s).

The Accumulator component is where metric events are concurrently
passed to an Aggregator, and it is this component that is most
responsible for SDK performance.  The Accumulator is responsible for
organizing bound and unbound instruments, updating and synchronized
copying of Aggregator state, invoking Observer instruments, and the
collection cycle.

The Processor component is intended as the most customizable component
in the export pipeline.  The Processor is responsible for selecting
Aggregators to use for specific instruments, via an independent
`AggregationSelector` interface, for reducing dimensionality, and for
conversion between DELTA and CUMULATIVE data point representation.
The Processor interface supports arbitrary protocol-independent data
transformation, and Processors can be chained together to form more
complex export pipelines.

The Exporter component converts processed data into a specific
protocol and exports it somewhere.  Following the [library
guidelines](../library-guidelines.md), exporters are expected to

> contain contain minimal functionality and customization is preferably

Reviewer 6: ```suggestion
contain minimal functionality and customization is preferably
```

expressed through Processors.

The Controller component coordinates the sequence of actions to
collect, process, and export metric data over a collection interval,
ensuring that access to the export pipeline is properly synchronized.

## Requirements

Requirements are listed below for the major components of the

> OpenTelemetry using the terminology outlined above.

Reviewer 6: right?

```suggestion
OpenTelemetry SDK using the terminology outlined above.
```

overall architecture for the SDK, just to name the stages of an export
pipeline and place them in context.

SDKs are required to use the standard names for these components, as
this helps build consistency across OpenTelemetry.  Every SDK SHOULD
include these components and the interfaces detailed below, although
the actual organization of each SDK may differ depending on the
available libraries and performance characteristics of the source
language.  For example, an SDK could be implemented using one
Accumulator per instrument or it could be implemented using one
Accumulator per collection period (assuming support for multiple
collection periods): these are considered implementation details.

### SDK

The SDK is the start an OpenTelemetry Metric export pipeline,
implementing the instrument-related APIs of the [`Meter`
interface](api.md#meter-interface) and providing the SDK instrument,
the Resource, and managing Instrumentation Library metadata.

An SDK is assembled with one or more Accumulators and other
components, detailed below.  The diagram below shows the relationship
between the API and the accumulator, with detail shown for synchronous

> instruments.

Reviewer 6: Did the diagram move? Is this is referring to the one in the Accumulator section?

Author: Fixed. I removed some of this text as off-topic, and introduced the diagram below.

and Accumulator; it is considered an implementation detail whether a
the SDK maintains one Accumulator, one Accumulator per instrument, or
some configuration in between.

The SDK `Collect()` function MUST call through to the Processor with
Accumulations from the active synchronous instruments as well as all
the registered asynchronous instruments.

The SDK MUST permit synchronous metric instruments to be used during
evaluation of asynchronous instrument callbacks.  The use of
synchronous instruments from asynchronous instrument callbacks is
considered a side-effect, in this case.  SDKs SHOULD process
synchronous instruments after asynchronous instruments, so that
side-effect synchronous measurements are processed as part of the same
collection interval that contains the corresponding asynchronous
observations.

### Accumulator

For a synchronous instrument, the accumulator will:

1. Map each active Label Set to a record, consisting of two instances of the same type Aggregator
2. Enter new records into the mapping, calling the AggregationSelector if needed
3. Update the current Aggregator instance, responding to concurrent API events

> 4. Call Aggregator.SynchronizedMove on the current Aggregator instance to copy its value into the snapshot Aggregator instance and reset itself

Author: Thanks. I had meant your second choice, "the current Aggregator".

I've updated this to read: "Call Aggregator.SynchronizedMove on the current Aggregator instance to: (a) copy its value into the snapshot Aggregator instance and (b) reset the current Aggregator to the zero state."

FWIW, the OTel-Go SDK documents this method with:

```
	// SynchronizedMove is called during collection to finish one
	// period of aggregation by atomically saving the
	// currently-updating state into the argument Aggregator AND
	// resetting the current value to the zero state.
	//
	// SynchronizedMove() is called concurrently with Update().  These
	// two methods must be synchronized with respect to each
	// other, for correctness.
	//
	// After saving a synchronized copy, the Aggregator can be converted
	// into one or more of the interfaces in the `aggregation` sub-package,
	// according to kind of Aggregator that was selected.
	//
	// This method will return an InconsistentAggregatorError if
	// this Aggregator cannot be copied into the destination due
	// to an incompatible type.
	//
	// This call has no Context argument because it is expected to
	// perform only computation.
	SynchronizedMove(destination Aggregator, descriptor *metric.Descriptor) error
```

Unless concurrency is not a feature of the source language, the SDK
Accumulator component SHOULD be designed with concurrent performance
in mind.

The Accumulator MAY use exclusive locking to maintain a map of
synchronous instrument updates.  The Accumulator SHOULD NOT hold an
exclusive lock while calling an Aggregator (see below), since some
Aggregators may have higher concurrency expectations.

#### Accumulator: Collect() function

The Accumulator MUST implement a Collect method that builds and
processes current Accumulation values for active instruments, meaning
those that were updated since the prior collection.  The Collect
method MUST call the Processor to process Accumulations corresponding
to all metric events that happened before the call.

Accumulations MUST be computed during Collect using a _synchronized
move_ operation on the Aggregator.  This operation, using some kind of synchronization, copies
the current Aggregator and resets it to the zero state, so that each
Aggregator immediately begins accumulating events for the next
collection period while the current one is processed.  An Accumulation
is defined as the synchronously-copied Aggregator combined with the

> LabelSet, Resource, and metric Descriptor.

Reviewer 6: For consistency:

```suggestion
Label Set, Resource, and metric Descriptor.
```

OpenTelemetry using the terminology outlined above.

The data flow diagram above is meant as an abstraction, for
understanding how the standard component names used in this document
map into a data path.  The diagram is not meant to prescribe an
overall architecture for the SDK, just to name the stages of an export
pipeline and place them in context.

SDKs are required to use the standard names for these components, as
this helps build consistency across OpenTelemetry.  Every SDK SHOULD
include these components and the interfaces detailed below, although
the actual organization of each SDK may differ depending on the
available libraries and performance characteristics of the source
language.  For example, an SDK could be implemented using one
Accumulator per instrument or it could be implemented using one
Accumulator per collection period (assuming support for multiple
collection periods): these are considered implementation details.

### Instrument Registration

The OpenTelemetry SDK MUST ensure that an individual Meter
implementation cannot report instruments with the same name and
different instrument kind.

For synchronous instruments, duplicate registration is permitted.  The

> SDK SHOULD return the same SDK instrument when correctly registered

Reviewer: The "same" SDK instrument is determined how? By name?

Reviewer: Maybe linking to [naming guidelines](https://github.com/open-telemetry/oteps/blob/master/text/metrics/0108-naming-guidelines.md) would be good here?

Author: This concept of supporting duplicate registration has mostly fallen apart, and I rewrote this section based on new thinking. The original idea came from OpenCensus, which did not have a named-Meter-like concept. When I think about the possible upside of supporting duplicate registrations, it only applies in a world without named Meters.

I now believe that if there is a single instrumentation library defining instruments, it _ought_ to be an error to define the same instrument name more than once, even with identical information. 

I've updated this section to say that duplicate registration is not permitted. I don't think this will create any problems, and it's far easier to specify. Do you agree @bogdandrutu?

Author: I added a link to the naming guidelines with a TODO. OTEP 108 should be written into the specification somewhere, and most of the guidelines are just guidelines. There will be other material written into the spec that applies to both Tracing and Metrics (and Logging, I assume) about character sets, label/attribute lengths, and so on. Those are requirements for the SDK as well, and should be linked here, but also do not exist in the current specification. (I believe the Tracing push to GA will address these if it hasn't already.)

multiple times.

For asynchronous instruments, duplicate registration is not permitted.
The SDK MUST not permit registration of multiple callbacks with the same
instrument name.

### Accumulator

The Accumulator is the first component an OpenTelemetry Metric export
pipeline, implementing the instrument-related APIs of the [`Meter`
interface](api.md#meter-interface) and providing the SDK instrument.
An SDK is assembled with an Accumulator and other parts, detailed
below.

The Accumulator MUST provide the option to associate a
[`Resource`](../resource/sdk.md) with the Accumulations that it
produces.

Synchronous metric instruments are expected to be used concurrently.
Unless concurrency is not a feature of the source language, the SDK
Accumulator component SHOULD be designed with concurrent performance
in mind.

The Accumulator MAY use exclusive locking to maintain a map of
synchronous instrument updates.  The Accumulator SHOULD NOT hold an
exclusive lock while calling an Aggregator (see below), since some
Aggregators may have higher concurrency expectations.

> 

Reviewer: Should there be a section on collection?

Author: I added one, see what you think.

Author: (I added that this now fixes #607, as a result.)

Reviewer: It would be great to have the `collect` definition outlined in the model implementation (Go example) as well.

Author: Done. There is a new diagram and a new header sections calling out Collect().

- **ExportKind**: one of Delta, Cumulative, or Pass-Through
- **ExportKindSelector**: chooses which ExportKind to use for a metric instrument.
- **ExportRecord**: consists of Instrument, Label Set, Resource, Timestamp(s), and Aggregation
- **ExportRecordSet**: a set of export records.

The term **SDK instrument** refers to the underlying implementation of
an instrument.

## Dataflow Diagram

From an external perspective, the Metrics SDK implements the `Meter`
and `Provider` interfaces described in the [Metrics API](api.md).
From an internal perspective, the Metrics SDK encapsulates an export
pipeline for metric data.

![Metrics SDK Design Diagram](img/metrics-sdk.png)

The Accumulator component is where metric events leave the API and
enter the export pipeline, and it is this component that is most
responsible for SDK performance.  The Accumulator is responsible for
organizing bound and unbound instruments, updating and synchronized
copying of Aggregator state, invoking Observer instruments, and the
collection cycle.

The Processor component is intended as the most customizable component
in the export pipeline.  The Processor is responsible for selecting

> Aggregators to use for specific instruments, reducing dimensionality,

Reviewer 7: Are `Processor`s responsible for selecting aggregators to use for instruments? I've seen statements above that `AggregatorSelector`s are responsible for that.

Reviewer 2: Bit confused here. From the architecture diagram, Accumulator receives metric events, and accumulates them (using Aggregator) and passed the Accumulation to the Processor. This means Accumulator should know which aggregator to use, and processor simply receives the Accumulations. Processor does not know anything about aggregators. -> Is this right understanding?

Author: I've added a diagram to clarify the intent, and added to the sentence here indicating that there is an independent sub-component called `AggregationSelector` that is chooses the `Aggregator`. Ultimately this is something configurable, a choice made while setting up the SDK, depending on what kinds of configuration are possible and what the exporter actually supports.

![Metrics SDK Design Diagram](img/metrics-sdk.png)

For the `Meter` and `MeterProvider` interfaces, the SDK itself is
required to manage the Resource and Instrumentation Library metadata,
support synchronous and asynchronous instrument constructors, act as
an instrument registry (detect conflicts), and implement `RecordBatch`
functionality.

To support the export pipeline internally, the SDK itself is required
to implement a top-level `Collect()` API that runs collection on one
or more Accumulator(s).  This specification deliberately avoids
requiring a specific relationship between the SDK and Accumulator, it
is considered an implementation detail whether a the SDK maintains one
Accumulator, one Accumulator per instrument, or some configuration in

> between.

Reviewer 3: 💯 👍

### Instrument Registration

The OpenTelemetry SDK is responsible for ensuring that an individual
Meter implementation cannot report multiple instruments with the same
name and different definitions.  To accomplish this, SDKs MUST reject
duplicate registration of an instrument when another instrument has
already been registered with same metric name to the same named Meter.

> Separate Meters, characterized by different instrumetation library

Reviewer: ```suggestion
Separate Meters, characterized by different instrumentation library
```

Author: Fixed.
